04-onion-routing.md

```diff
@@ -150,52 +155,75 @@ The overall structure of the packet is as follows:
 
 1. type: `hops_data`
 2. data:
-   * [`1`:`realm`]
-   * [`32`:`per_hop`]
-   * [`32`:`HMAC`]
+   * [`n1*FRAME_SIZE`: `hop_payload`]
+   * [`n2*FRAME_SIZE`: `hop_payload`]
    * ...
    * `filler`
 
-Where, the `realm`, `per_hop` (with contents dependent on `realm`), and `HMAC`
-are repeated for each hop; and where, `filler` consists of obfuscated,
-deterministically-generated padding, as detailed in
-[Filler Generation](#filler-generation). Additionally, `hops_data` is
-incrementally obfuscated at each hop.
+Where `filler` consists of obfuscated, deterministically-generated padding, as detailed in [Filler Generation](#filler-generation).
+Additionally, `hop_payload`s are incrementally obfuscated at each hop.
 
-The `realm` byte determines the format of the `per_hop` field; currently, only `realm`
-0 is defined, for which the `per_hop` format follows:
+Each `hop_payload` has the following structure:
 
-1. type: `per_hop` (for `realm` 0)
+1. type: `hop_payload`
+2. data:
+   * [`1`: `num_frames_and_realm`]
+   * [`raw_payload_len`: `raw_payload`]
+   * [`padding_len`: `padding`]
+   * [`32`: `HMAC`]
+
+Notice that since the `hop_payload` is instantiated once per hop, the subscript `_i` may be used in the remainder of this document to refer to the `hop_payload` and its fields destined for hop `i`.
+
+The `hop_payload` consists of at least one `frame` followed by up to 15 additional `frame`s.
+The number of additional frames allocated to the current hop is determined by the 4 most significant bits of `num_frames_and_realm`, while the 4 least significant bits determine the payload format.
+Therefore the number of frames allocated to the current hop is given by `num_frames = (num_frames_and_realm >> 4) + 1`.
+For simplification we will use `hop_payload_len` to refer to `num_frames * FRAME_SIZE`.
+
+In order to have sufficient space to serialized the `raw_payload` into the `hop_payload` while minimizing the number of used frames the number of frames used for a single `hop_payload` MUST be equal to
+
+> `num_frames = ceil((raw_payload_len + 1 + 32) / FRAME_SIZE)`
+
+The payload format determines how the `raw_payload` should be interpreted (see below for currently defined formats), and how much padding is added.
+In order to position the `HMAC` in the last 32 bytes of the `hop` the `raw_payload` MUST be followed by `padding_len = (num_frames * FRAME_SIZE - 1 - raw_payload_len - 32)` `0x00`-bytes.
+
+The `realm` is specified as `num_frames_and_realm & 0x0F`.
+It determines the format of the `raw_payload` field; the following `realm`s are currently defined.
+
+| `realm` | Payload Format                                             |
+|:--------|:-----------------------------------------------------------|
+| `0x00`  | [Version 1 `hop_data`](#version-1-hop_data-payload-format) |
+| `0x01`  | TLV payload format (to be specified)                       |
```

Comment:
Let's call this "TLV Version 1"? Just to make it clear that this is only a single TLV namespace and doesn't apply globally (as that would be too limiting IMO).

Reply:
I don't quite get the differentiation.

Didn't we discuss yesterday that the TLV types should be message specific? In other words we'd be adding a section below listing all the possible TLV types and their meaning. Version 1 seems to indicate that there are multiple versions the TLV serialization format, whereas we use the TLV spec for details on how to encode/decode TLV entries, and then use the table below to decipher them.

Reply:
@Roasbeef correct me if I'm wrong, but here is how I understood your comment (which I agree with).

This should be called "TLV Version 1" because this will encompass the TLV messages used with `realm=0x01`. As @cdecker points out, all the types and meaning for that TLV Version 1 would then be documented in a section below.

But in the future, we might want to create another realm (`0x02` for example) that would also use TLV, but with different tags. These new TLV types would be documented in their own section in this spec, but they could completely ignore the TLV types of realm 0x01.

Does it make sense? Was that what you meant?

Reply:
So the realm and packet can be seen as a 2x2 matrix, which allows us to extend the total number of possible distinct K-V pairs for the TLV. It's not that there're multiple versions of the format for the onion (I still think we need distinct TLV for onion and wire messages, but that's another topic), but that there're multiple TLV _namespaces_. I'm worried that will all the developer excitement around this new feature, and all the uses we've _already_ thought of, 255 K-V pairs will be far from sufficient. Introducing TLV namespaces for the onion allows us to extend that further, and even further yet using the packet type to create nested namspaces. Packet type `x00` being the first namespace.

Reply:
If we extend this to be per frame (leave realm as is, use 2 bytes within a frame), then it allows a sender to drop information from _distinct_ namespaces to a particular hop.